kolibri/core/tasks/test/test_api.py

```diff
@@ -118,6 +118,18 @@ def assert_clearable(index, expected):
             for i in [4, 5, 6]:
                 assert_clearable(i, True)
 
+    def test_restart_task(self, queue_mock):
+        queue_mock.restart_job.return_value = 2
+        queue_mock.restart_job.return_value = fake_job(state=State.QUEUED)
```

Comment:
Pass in `id=2` here, and get rid of the line above. That will force your fake job to have an id of 2, so your assertion at the end of the test should now be true.

Reply:
I tried that. Still have the error.
There are two queue functions that I need to `patch` here - 
1. `restart_job`
2. `fetch_job`

Here is how they are used in the implementation `restarttask` endpoint in the api - 
```
            try:
                task_id = _queue.restart_job(request.data["task_id"])
                resp = _job_to_response(_queue.fetch_job(task_id))
                break
```

`restart_job` should return an integer with value 2
`fetch_job` should return a fake job with State=Queue, id=2

Error is at the stage where `restart_job` raises `JobNotFound`. Hence it isn't being patched.

Reply:
Oh, sorry, I understand what you are trying to do now.

In that case, I think the only error here is that the second line should be:
```
queue_mock.fetch_job.return_value = fake_job(state=State.QUEUED, id=2)
```

Because at the moment you are only mocking `restart_job` and not `fetch_job` too.

Reply:
Oh my bad for that. But it still fails even after updating it to - 
```
queue_mock.restart_job.return_value = 2
queue_mock.fetch_job.return_value = fake_job(state=State.QUEUED, id=2)
```

Reply:
Ah yes, you are right - I just checked this out locally. The issue here is that the `@patch("kolibri.core.tasks.api.queue")` decorator on the TestCase does not seem to be mocking anything - the other tests in this test case all pass, as they all pass in the case that a job is not found.

I think the issue is due to the fact that the TaskViewset `queues` property is set to a list of `[queue, priority_queue]` because this is defined at module parse time, the mocking fails to intercept this, because at the time that `[queue, priority_queue]` is defined the references are to the original, non-mocked objects.

The easiest way to make this work would probably be to turn the `queues` property of the TaskViewset into:
```
@property
def queues(self):
    return [queue, priority_queue]
```

This change along with passing `job_id=2` instead of `id=2` (my mistake), causes the tests to now pass.

Reply:
I see. This was something new to me. Thank you for the help @rtibbles!
The test seems to be fixed.

Reply:
Yeah, understanding when and how mocks actually apply can be a little bit difficult to wrap one's head around sometimes!